algorithms/src/polycommit/sonic_pc/mod.rs

```diff
@@ -338,91 +306,102 @@ impl<E: PairingEngine> PolynomialCommitment<E::Fr, E::Fq> for SonicKZG10<E> {
         Ok((labeled_comms, randomness))
     }
 
-    fn open<'a>(
-        ck: &Self::CommitterKey,
+    pub fn combine_for_open<'a>(
+        ck: &CommitterKey<E>,
         labeled_polynomials: impl IntoIterator<Item = &'a LabeledPolynomial<E::Fr>>,
-        _commitments: impl IntoIterator<Item = &'a LabeledCommitment<Self::Commitment>>,
-        point: E::Fr,
-        opening_challenge: E::Fr,
-        rands: impl IntoIterator<Item = &'a Self::Randomness>,
-        _rng: Option<&mut dyn RngCore>,
-    ) -> Result<Self::Proof, PCError>
+        rands: impl IntoIterator<Item = &'a Randomness<E>>,
+        fs_rng: &mut S,
+    ) -> Result<(DensePolynomial<E::Fr>, Randomness<E>), PCError>
     where
-        Self::Randomness: 'a,
-        Self::Commitment: 'a,
+        Randomness<E>: 'a,
+        Commitment<E>: 'a,
     {
-        let mut combined_polynomial = DensePolynomial::zero();
-        let mut combined_rand = kzg10::Randomness::empty();
-        let mut curr_challenge = opening_challenge;
-
-        for (polynomial, rand) in labeled_polynomials.into_iter().zip_eq(rands) {
+        Ok(Self::combine_polynomials(labeled_polynomials.into_iter().zip_eq(rands).map(|(p, r)| {
             let enforced_degree_bounds: Option<&[usize]> = ck.enforced_degree_bounds.as_deref();
 
             kzg10::KZG10::<E>::check_degrees_and_bounds(
                 ck.supported_degree(),
                 ck.max_degree,
                 enforced_degree_bounds,
-                polynomial,
-            )?;
+                p,
+            )
+            .unwrap();
+            let challenge = fs_rng.squeeze_short_nonnative_field_element().unwrap();
+            (challenge, p.polynomial().as_dense().unwrap(), r)
+        })))
+    }
 
-            combined_polynomial += (curr_challenge, polynomial.polynomial());
-            combined_rand += (curr_challenge, rand);
-            curr_challenge *= &opening_challenge;
+    /// On input a list of labeled polynomials and a query set, `open` outputs a proof of evaluation
+    /// of the polynomials at the points in the query set.
+    pub fn batch_open<'a>(
+        ck: &CommitterKey<E>,
+        labeled_polynomials: impl IntoIterator<Item = &'a LabeledPolynomial<E::Fr>>,
+        commitments: impl IntoIterator<Item = &'a LabeledCommitment<Commitment<E>>>,
+        query_set: &QuerySet<E::Fr>,
+        rands: impl IntoIterator<Item = &'a Randomness<E>>,
+        fs_rng: &mut S,
+    ) -> Result<BatchProof<E>, PCError>
+    where
+        Randomness<E>: 'a,
+        Commitment<E>: 'a,
+    {
+        let poly_rand_comm: BTreeMap<_, _> = labeled_polynomials
```

Comment:
same as before: if the collection is not iterated over (or a specific order is expected for any other reason), a `HashMap` will be faster (~`O(1)` as opposed to ~`O(log(n))`) for lookup purposes

Reply:
(Same comment applies here too)

Reply:
what about the number of iterations in the enclosing loop? is it also expected to be small?